Working controller support

brad1111 opened the pull request:
Replaces the fake cursor with XY navigation.

# Changes:
- Now reports as Xbox in the useragent when on Xbox, this means Jellyfin-web will default to TV layout which is required for Jellyfin-web's gamepad support.
- Force enables gamepad support in Jellyfin.
- Works around a bug in WebView2 where it doesn't won't programatically focus the web content.
- Disables back button handling as Jellyfin-web handles this itself.

# Known issues:
- Left stick movement moves two spaces instead of one:
This is because WebView2 on Xbox only for some reason sends Arrow keydown events on left stick movement instead of KeyCodes 211-214. This means joystick left ends up causing a keydown event in keyboardNavigation.js, and then gamepadtokey.js fires a second one. This needs to be resolved in Jellyfin-web.
- Some face button inputs are processed twice.
Not entirely sure why this happens, but needs to be fixed from Jellyfin-web side.
- Toggling a checkbox untoggles itself
This is a bug in Jellyfin-web and also happens in other browers when using a controller. Essentially gamepadtokey.js creates a keydown event with 13, and then clicks the element on keyup.

I'm also planning to submit a some PRs to fix the Jellyfin web side. I'm suspecting we should probably bypass gamepadtokey.js altogether as a WebView on UWP provides javascript keydown events similar to how keyboard and tv remote support already works.

TimGels commented on Jellyfin/Controls/JellyfinWebView.xaml.cs:
```diff
@@ -31,16 +33,32 @@ public JellyfinWebView()
+            await WView_NavigationStartingTask(sender, args);
+        }
+
+        private void HandleGamepadBackPress()
```
Might as well remove this event listener then if it is redundant.

TimGels replied:
Good question. You can keep that file in the codebase. If it stays unused in the future, or gets superseeded by something else, we can always remove it when the time is right.

TimGels commented:
> Some face button inputs are processed twice.

Actually processed or send twice? If the latter we might need to debounce the input press.

TimGels commented:
> You are correct, I originally explicitly focused the WebView again, but I thought it was redundant. If it makes the code clearer I can add it back.

I don't find it the end of the world as long as the working of this work-around is explained in the code. But I would personally prefer explicit focus in case the XAML is changed in the future, possibly changing the focus priority on that page. Which could result in confusion.

thornbill commented:
Enabling the gamepad should be done in web in my opinion. Just change the default `false` value on this line to `!!browser.xboxOne`.

https://github.com/jellyfin/jellyfin-web/blob/master/src/scripts/settings/appSettings.js#L32

brad1111 commented:
> Enabling the gamepad should be done in web in my opinion. Just change the default `false` value on this line to `!!browser.xboxOne`.

I think the right solution is going to be to disable jellyfins gamepad support on Xbox and to add keycodes to https://github.com/jellyfin/jellyfin-web/blob/master/src/scripts/keyboardNavigation.js#L14 since native gamepad keydown events can't be turned off in WebView2, and gamepadtokey also creates duplicate keydown events. The alternative needs to patch a lot of different functions to ignore keypresses coming from the browser.



> @thornbill Xbox is no longer part of the UserAgent with WebView2 so `browser.xboxOne` is always false.
> 
> I have added Xbox (The deviceForm) to the useragent in https://github.com/jellyfin/jellyfin-uwp/pull/44/files#diff-6a5c95e26bac22e1eb36ae874d1f1d3451598e4b1db232f0ee346242509781baR73 but that hasn't been merged yet.

I also added the useragent for this PR though yours is much more comprehensive than mine. I simply just tack "WebView2 Xbox" onto the end (or whatever platform its being run on if it isn't Xbox).

TimGels commented:
@brad1111 

Sounds awesome and thanks for contributing!

Let me know when this ready for testing. I will then give this a test with a controller on my computer. I need to try and find some time during a weekend for this, but I will let you know if there are any findings.

JPVenson commented:
i will give it a final look tmrw because my dev VM exploded again ... fucking windows updates

brad1111 commented:
@TimGels 
> That sounds like a bug. If there is no issue for it yet, would you be open to creating an issue for it in [jellyfin-web](https://github.com/jellyfin/jellyfin-web) repository?

I have already created a PR and corresponding issue for this, though forgot to link it here: https://github.com/jellyfin/jellyfin-web/pull/7031

> So to make sure we are on the same page. You want to use nativeshell to remove the option for a user to disable tv mode? I personally don't see that as blocking for this pull-request. That can be resolved in another PR.

Correct, as that was suggested in https://github.com/jellyfin/jellyfin-web/pull/6993#discussion_r2175488170.

JPVenson commented:
I just tested your branch on my Xbox One with the controller and no button works for selecting anything. I logged into the server by using a keyboard but after that i am not able to select any profile or select any buttons

JPVenson reviewed (changes requested):
test failed

JPVenson commented:
Ok yea i tested with RC2 ... if we introduce some breaking change that requires 10.11 in the future i would need to wait for @TimGels to finish his connection refactor and then read the JF version to provide a meaningful error message

brad1111 commented:
Currently the check only works if they're connecting to a new server. In the scenario that someones updates their Jellyfin Xbox, but not their Jellyfin Server to 10.11, then no message would appear and they would be unable to navigate.

I'm currently quickly testing to see if we can re-enable mouse cursor, for older versions assuming there's no other breaking changes.
